test/unit/BPool/BPool_SwapExactAmountIn.tree

```diff
@@ -0,0 +1,26 @@
+BPool::SwapExactAmountIn
+├── when reentrancy lock is set
+│   └── it should revert
+├── when pool is not finalized
+│   └── it should revert
+├── when token in is not bound
+│   └── it should revert
+├── when token out is not bound
+│   └── it should revert
+├── when token in exceeds max allowed ratio
+│   └── it should revert
+├── when spot price before swap exceeds maxPrice
+│   └── it should revert
+├── when spot price after swap exceeds maxPrice
+│   └── it should revert
+├── when returned tokenOut is less than minAmountOut
+│   └── it should revert
+└── when preconditions are met
```

Comment:
ooops, realized I'm not triggering `BPool_SpotPriceBeforeAboveTokenRatio`. Looking into it

Reply:
the reason behind that check and the `MAX_{IN,OUT}_RATIO` ones sure looks like could be fun properties :eyes:

Reply:
i think the rationale is avoiding fat fingers, meaning buying more than 50% of the available tokens, or selling more than 50% of the pool balance.

as for the `BPool_SpotPriceBeforeAboveTokenRatio` isn't it deprecated in dev?

Reply:
I tried to find parameters to trigger this particular error and wasn't able to: https://www.desmos.com/calculator/onhfxbfljk

@drgorillamd you think it would be possible to formally prove this is unreachable?

I did notice that there is one testcase for `swapExactAmountOut`, will look into it when testing that method

Reply:
to be clear, we're not planning on removing the check even if we find it unreachable, but we can then skip the test, or add a comment in the test explaining the current behaviour

Reply:
> I tried to find parameters to trigger this particular error and wasn't able to: https://www.desmos.com/calculator/onhfxbfljk
> 
> @drgorillamd you think it would be possible to formally prove this is unreachable?
> 
> I did notice that there is one testcase for `swapExactAmountOut`, will look into it when testing that method

Quickly tried with Halmos but there are some loops hiding in bnum, so I'll try with Kontrol instead (Echidna seemds to find some cases tho, for small token amount in esp, will double-check tomorrow along Kontrol tho)